### PR TITLE
FP8 enablement - add a pseudorandom number generator, add conversion methods

### PR DESCRIPTION
**Pull request by geyyer**
Add a PRNG

### CODE REVIEW

**Comment by geyyer** [comment]
Add fp8<->float conversion with standard rounding.

**Comment by geyyer** [comment]
This PR adds:

- fp8 type definition
- pseudorandom number generator - the rocBLAS implementation
- conversion functions fp8<->fp32 with standard rounding
- conversion functions fp8<->fp32 with stochastic rounding
- conversion functions fp8<->fp16 with standard rounding
- conversion functions fp8<->fp16 with stochastic rounding
- elementwise ops for fp8<->fp32, fp8<->fp16 with standard rounding (PassThrough)
- elementwise ops for fp8<->fp32, fp8<->fp16 with stochastic rounding (ConvertF8SR)
- tests for both conversions using both rounding modes

File: include/ck/utility/f8_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+namespace ck {
+
+using f8_t   = uint8_t;
+using half_t = _Float16;
+
+// fp8 rounding modes
+enum class f8_rounding_mode
+{
+    standard,
+    stochastic
```
Comment (aosewski) [inline include/ck/utility/f8_utils.hpp:16]: What does it mean `standard` and what `stochastic` ? Could you please add documentation?

File: include/ck/utility/f8_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        int sh = 1 + __builtin_clz(mantissa) - ((1 + type_exp + type_mant) - f8_mant);
+        mantissa <<= sh;
+        exponent += 1 - sh;
+        /*
+        exponent++;
+        while(mantissa<(1<<wm)) {
+        mantissa <<= 1;
+        exponent--;
+        }
+        */
```
Comment (aosewski) [inline include/ck/utility/f8_utils.hpp]: Leftover code?

File: include/ck/utility/f8_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#pragma once
+
+#include "ck/utility/statically_indexed_array.hpp"
+
+namespace ck {
```
Comment (aosewski) [inline include/ck/utility/f8_utils.hpp:8]: If below functions are utilities I'd put them under `ck::utils` namespace exposing only API and hiding implementation functions (aka `run...`) under `ck::utils::detail::unnamed namespace` in this file. This way you limit the number of symbols exposed to the user and hide implementation details.

File: include/ck/utility/data_type.hpp [hunk 4]
```diff
@@ -1090,6 +1145,82 @@ inline __host__ __device__ constexpr bhalf_t bf16_convert_rtn<bhalf_t, half_t>(h
...
+{
+    constexpr bool negative_zero_nan = true;
+    constexpr bool clip              = true;
+    constexpr f8_rounding_mode rm    = f8_rounding_mode::stochastic;
+    constexpr int seed               = 42;
+    // as thread id is not available on host, use 0 for prn generation
+    uint32_t rng = prand_generator<half_t, seed>(reinterpret_cast<uintptr_t>(&x), x);
+    return cast_to_f8<half_t, negative_zero_nan, clip, (rm == f8_rounding_mode::stochastic)>(x,
+                                                                                             rng);
+}
```
Comment (aosewski) [inline include/ck/utility/data_type.hpp]: I see that the number of `type_convert`-like functions is increasing significantly. Maybe we should consider separating them into the file `type_convert.hpp` ?

File: include/ck/utility/data_type.hpp [hunk 5]
```diff
@@ -1090,6 +1145,82 @@ inline __host__ __device__ constexpr bhalf_t bf16_convert_rtn<bhalf_t, half_t>(h
...
+          uint32_t seed_t,
+          std::enable_if_t<!(std::is_same<float, T>{} || std::is_same<half_t, T>{}), bool> = false>
+__host__ __device__ uint32_t prand_generator(int id, T val, uint32_t seed = seed_t)
+{
+    std::ignore = id;
+    std::ignore = val;
+    std::ignore = seed;
+
+    return 0;
+}
```
Comment (aosewski) [inline include/ck/utility/data_type.hpp]: These are implementation detail, helper functions. Maybe we should put them into separate file?

**Review by aosewski — COMMENTED** [review]
Looks very good. I just have few observations regarding overall code structure. I'm not familiar with those bit-wise algorithms so I'm not even touching it :P

File: include/ck/utility/f8_utils.hpp (hunk 1 above)
Comment (geyyer) [reply include/ck/utility/f8_utils.hpp:16]: Added in [f730c3f](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/708/commits/f730c3fb505571df42a8f664ec23bcec01a96425)

File: include/ck/utility/f8_utils.hpp (hunk 2 above)
Comment (geyyer) [reply include/ck/utility/f8_utils.hpp]: Removed in [f61c770](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/708/commits/f61c77040aa8e62bb14134925016c54682b9263e)

File: include/ck/utility/f8_utils.hpp (hunk 3 above)
Comment (geyyer) [reply include/ck/utility/f8_utils.hpp:8]: Do you mean something like [562ec12](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/708/commits/562ec121101e28b9301be9f2a0c2f2171652984b)?

File: include/ck/utility/data_type.hpp (hunk 4 above)
Comment (geyyer) [reply include/ck/utility/data_type.hpp]: Added a header in [d6a666f](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/708/commits/d6a666fa389b14c7ef1abc0f473d11ffd0c4062c)

File: include/ck/utility/data_type.hpp (hunk 5 above)
Comment (geyyer) [reply include/ck/utility/data_type.hpp]: Added another header in [c208a8a](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/708/commits/c208a8acea1a11b9f58a9768ef1bc4749be7fcc4)

**Comment by geyyer** [comment]
Casting functions are still far from optimal, I'll work on their performance as we have a test case to track conversion performance.

File: include/ck/utility/type_convert.hpp [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+#include "ck/utility/f8_utils.hpp"
+#include "ck/utility/random_gen.hpp"
+
+namespace ck {
+
+using bhalf_t = ushort;
+using half_t  = _Float16;
+#ifdef CK_EXPERIMENTAL_BIT_INT_EXTENSION_INT4
+using int4_t = _BitInt(4);
+#endif
```
Comment (aosewski) [inline include/ck/utility/type_convert.hpp]: I'm not sure if this is a good solution to duplicate here these type aliases. The `type_convert.hpp`'s file content is inherently connected with the `data_type.hpp` header, so I'd rather `#include` it here. Moreover, then we don't risk to maintain multiple type aliases at different places.

File: include/ck/utility/data_type.hpp [hunk 7]
```diff
@@ -4,6 +4,7 @@
 #pragma once
 
 #include "ck/utility/statically_indexed_array.hpp"
+#include "ck/utility/type_convert.hpp"
```
Comment (aosewski) [inline include/ck/utility/data_type.hpp]: I think that this relation is the other way round. The `type_convert.hpp` should depend on data types defined in `data_type.hpp` header. In this file there shouldn't be anything which relies on types/functions from `type_convert.hpp`.

**Review by aosewski — COMMENTED** [review]
Thanks for refactoring files! The only one thing which I don't feel is OK is the relation between `type_convert.hpp` and `data_type.hpp` headers. Please see comments,

File: include/ck/utility/type_convert.hpp (hunk 6 above)
Comment (geyyer) [reply include/ck/utility/type_convert.hpp]: Removed in [08e263e](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/708/commits/08e263e348b16202dfcb0f2605e373510e1e950f)

File: include/ck/utility/data_type.hpp (hunk 7 above)
Comment (geyyer) [reply include/ck/utility/data_type.hpp]: Updated dependencies in [08e263e](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/708/commits/08e263e348b16202dfcb0f2605e373510e1e950f)

**Comment by geyyer** [comment]
@aosewski, thanks for your review, I've refactored the type_convert.hpp, could you review the change?

File: include/ck/utility/f8_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,251 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#pragma once
+
+namespace ck {
+
+using f8_t   = uint8_t;
+using half_t = _Float16;
```
Comment (aosewski) [inline include/ck/utility/f8_utils.hpp]: Oh I've overlooked this. Please move the fp8 type alias to `data_type.hpp` and remove the `half_t` type alias. As I said for such simple type aliases definitions I think we should keep all the data type definitions in `data_type.hpp` header. I imagine an exception to this rule only when we would define our own class for a specific data type.

**Review by aosewski — DISMISSED** [review]
I've overlooked the place of `f8_t`  type definition. Other than that looks good. I do not want to prolong the review process, thus I approve. But please move this type definition and everything should be fine then. Thanks!

File: include/ck/utility/f8_utils.hpp (hunk 8 above)
Comment (geyyer) [reply include/ck/utility/f8_utils.hpp]: Fixed in [2ee1c0a](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/708/commits/2ee1c0a70a305b00770c51a6ad585895992b327c)

**Comment by geyyer** [comment]
@aosewski, looks like I've overridden your approval, could you approve one more time?

**Review by aosewski — APPROVED** [review]

**Pull request merged**